### ISSUE TITLE
Lazy modules causes shared module to be instanciated twice

### ISSUE THREAD

**Issue by mistic100** [angular/angular#30150]
# 🐞 bug report

### Description

When sharing a single module accross multiple lazy loaded modules, the constructor of the said shared module is called once by lazy module + root module.

This makes impractical to have a shared module having initialization steps, like dispatching an NgRx Action to check the user authentication, load server-side configuration, etc. Because the Action is dispatched multiple times.

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-jwr53t

Open the console, "instanciate shared module" will be logged once at app startup and once more on each lazy module (Customers and Orders)

## 🌍  Your Environment

**Angular Version:**
<pre>
Angular CLI: 7.3.8
Node: 8.12.0
OS: linux x64
Angular: 7.2.14
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.13.8
@angular-devkit/build-angular      0.13.8
@angular-devkit/build-ng-packagr   0.13.8
@angular-devkit/build-optimizer    0.13.8
@angular-devkit/build-webpack      0.13.8
@angular-devkit/core               7.3.8
@angular-devkit/schematics         7.3.8
@angular/cli                       7.3.8
@ngtools/json-schema               1.1.0
@ngtools/webpack                   7.3.8
@schematics/angular                7.3.8
@schematics/update                 0.13.8
ng-packagr                         4.7.1
rxjs                               6.4.0
typescript                         3.2.4
webpack                            4.29.0
</pre>

**Comment by alxhub** [comment]
tl;dr: Don't put initialization logic in shared modules.

This is working as expected. In Angular, `NgModule`s are instantiated (and thus any initialization logic within them is run) whenever the module is included in an injector. So a if a `SharedModule` is used across the app, then it will be instantiated multiple times:

* once, for the application root injector
* again if it's included in any lazily loaded module, as each lazily loaded module has its own injector

Semantically it doesn't really make sense for initialization logic to be in a shared module - put it in a separate module which is only included in the app root.

**Comment by mistic100** [comment]
@alxhub ok, what is the recommended way for global and automatic initialization logic ?

It is an enterprise shared module, developers must not be able to disable this init, or forgot to call it.
